pkg/planner/core/find_best_task.go

```diff
@@ -711,34 +710,71 @@ func compareGlobalIndex(lhs, rhs *candidatePath) int {
+		if !statsTbl.ColAndIdxExistenceMap.HasAnalyzed(lhs.path.Index.ID, true) {
+			idxMissingStats, lhsPseudo = true, true
+		}
 	}
-	rhsHasStatistics := statsTbl.Pseudo
+	rhsPseudo := statsTbl.HistColl.Pseudo
 	if statsTbl != nil && rhs.path.Index != nil {
-		rhsHasStatistics = statsTbl.ColAndIdxExistenceMap.HasAnalyzed(rhs.path.Index.ID, true)
+		if !statsTbl.ColAndIdxExistenceMap.HasAnalyzed(rhs.path.Index.ID, true) {
+			idxMissingStats, rhsPseudo = true, true
+		}
 	}
-	if !lhs.path.IsTablePath() && !rhs.path.IsTablePath() && // Not a table scan
-		(lhsHasStatistics || rhsHasStatistics) && // At least one index has statistics
-		(!lhsHasStatistics || !rhsHasStatistics) && // At least one index doesn't have statistics
+	// It's either "not a table scan" or it's a table scan that scans a subset of the PK
+	lhsLimitedScan := !lhs.path.IsTablePath() || (lhs.path.IsTablePath() && !lhs.path.IsFullTableRange(tableInfo))
+	rhsLimitedScan := !rhs.path.IsTablePath() || (rhs.path.IsTablePath() && !rhs.path.IsFullTableRange(tableInfo))
+	// First set of rules apply to situations where an index doesn't have statistics
+	if (lhsLimitedScan || rhsLimitedScan) && // Not a full table scan
+		(!lhsPseudo || !rhsPseudo || !statsTbl.HistColl.Pseudo) && // An index or the table has statistics
+		idxMissingStats && // At least one index doesn't have statistics
 		len(lhs.path.PartialIndexPaths) == 0 && len(rhs.path.PartialIndexPaths) == 0 { // not IndexMerge due to unreliability
-		lhsTotalEqual := lhs.path.EqCondCount + lhs.path.EqOrInCondCount
-		rhsTotalEqual := rhs.path.EqCondCount + rhs.path.EqOrInCondCount
-		if lhsHasStatistics && lhsTotalEqual > 0 && lhsTotalEqual >= rhsTotalEqual {
-			return 1
-		}
-		if rhsHasStatistics && rhsTotalEqual > 0 && rhsTotalEqual >= lhsTotalEqual {
-			return -1
+		if lhs.path.EqOrInCondCount > 0 || rhs.path.EqOrInCondCount > 0 {
+			// If one index has statistics and the other does not, choose the index with statistics if it
+			// has the same or higher number of equal/IN predicates.
+			// This is to limit a new index being chosen until it's statistics are collected if we cannot determine
+			// that it is "potentially" more selective.
+			if lhsLimitedScan && !lhsPseudo && lhs.path.EqOrInCondCount >= rhs.path.EqOrInCondCount {
+				return 1, lhsPseudo
+			}
+			if rhsLimitedScan && !rhsPseudo && rhs.path.EqOrInCondCount >= lhs.path.EqOrInCondCount {
+				return -1, rhsPseudo
+			}
+			if preferRange {
+				// keep an index without statistics if that index has more equal/IN predicates, AND:
+				// 1) there first must be at least a 5% change in estimated number of rows, AND
+				// 2) there are at least 2 equal/INs, OR - when combined with 1) above
+				// 3) it's a full index match for all index predicates
+				planCountDiff := lhs.path.CountAfterAccess / rhs.path.CountAfterAccess
+				if lhs.path.CountAfterAccess <= 1 || planCountDiff < 0.95 || planCountDiff > 1.05 {
```

Comment:
Why use `lhs.path.CountAfterAccess <= 1` here? It seems like `planCountDiff < 95% or planCountDiff > 105%` is enough to implement the rule (1): `there first must be at least a 5% change in estimated number of rows`

Reply:
And another question is that since some stats is missing, do we still have to consider the estimated number of rows in this case? To me, only rule (2) and (3) seems enough.